### ISSUE TITLE
'Juju bootstrap microk8s' is not working in offline machine

### ISSUE THREAD

**Issue by morenoloza98** [canonical/microk8s#3956]
#### Summary
I am trying to reproduce an installation environment where there is no internet access for the machine. It is worth mentioning that it is a Rocky Linux 8 machine and juju and microk8s are being installed using snap. 
I installed snap, microk8s and juju by downloading the packages and transferring them to the offline machine. Addons for microk8s are already installed as well. The first issue I encountered was when trying to enable the ingress addon for microk8s, as I get the following error:

```
sudo -u user /var/lib/snapd/snap/bin/microk8s enable ingress
Infer repository core for addon ingress
Enabling Ingress
E0504 09:57:46.134078  101790 memcache.go:287] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0504 09:57:46.150457  101790 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
ingressclass.networking.k8s.io/public unchanged
Warning: Detected changes to resource ingress which is currently being deleted.
namespace/ingress unchanged
clusterrole.rbac.authorization.k8s.io/nginx-ingress-microk8s-clusterrole unchanged
clusterrolebinding.rbac.authorization.k8s.io/nginx-ingress-microk8s unchanged
Error from server (Forbidden): error when creating "/home/user/snap/microk8s/4959/tmp/temp.yaml": serviceaccounts "nginx-ingress-microk8s-serviceaccount" is forbidden: unable to create new content in namespace ingress because it is being terminated
Error from server (Forbidden): error when creating "/home/user/snap/microk8s/4959/tmp/temp.yaml": roles.rbac.authorization.k8s.io "nginx-ingress-microk8s-role" is forbidden: unable to create new content in namespace ingress because it is being terminated
Error from server (Forbidden): error when creating "/home/user/snap/microk8s/4959/tmp/temp.yaml": rolebindings.rbac.authorization.k8s.io "nginx-ingress-microk8s" is forbidden: unable to create new content in namespace ingress because it is being terminated
Error from server (Forbidden): error when creating "/home/user/snap/microk8s/4959/tmp/temp.yaml": configmaps "nginx-load-balancer-microk8s-conf" is forbidden: unable to create new content in namespace ingress because it is being terminated
Error from server (Forbidden): error when creating "/home/user/snap/microk8s/4959/tmp/temp.yaml": configmaps "nginx-ingress-tcp-microk8s-conf" is forbidden: unable to create new content in namespace ingress because it is being terminated
Error from server (Forbidden): error when creating "/home/user/snap/microk8s/4959/tmp/temp.yaml": configmaps "nginx-ingress-udp-microk8s-conf" is forbidden: unable to create new content in namespace ingress because it is being terminated
Error from server (Forbidden): error when creating "/home/user/snap/microk8s/4959/tmp/temp.yaml": daemonsets.apps "nginx-ingress-microk8s-controller" is forbidden: unable to create new content in namespace ingress because it is being terminated
```

When I try to run ```juju bootstrap microk8s``` I set the flag --metadata-source to pass a directory that contains the following files:

- streams/v1/
- --- index.json
- --- com.ubuntu.cloud-released-imagemetadata.json
- public-clouds.yaml
- juju-metadata
- juju
- index2.json
- index.json

This is the output after running the command:

``` 
juju bootstrap microk8s --metadata-source /directory/to/images
Creating Juju controller "microk8s-localhost" on microk8s/localhost
Bootstrap to Kubernetes cluster identified as microk8s/localhost
Creating k8s resources for controller "controller-microk8s-localhost"
ERROR failed to bootstrap model: creating controller stack: creating statefulset for controller: timed out waiting for controller pod: pending:  - 
```

The controller can be seen in the list if juju controllers is run, but if I try running anything else, it throws the following error:

ERROR no controller API addresses; is bootstrap still in progress?

#### What Should Happen Instead?
What should happen instead is that the controller should be created so kubeflow and mlmd can be deployed for the app I am trying to run inside the offline machine.

#### Reproduction Steps
To reproduce the issue, you will need a Rocky Linux 8 machine with no internet access.

1. Download the necessary files and transfer them to the offline machine. Here is a list of files used:
- epel-release-8-18.el8.noarch.rpm
- microk8s-core-addons
<img width="240" alt="image" src="https://user-images.githubusercontent.com/45719935/236279130-6326ca5e-655d-4c47-9d1c-f6c35c60fcdc.png">
<img width="98" alt="image" src="https://user-images.githubusercontent.com/45719935/236279446-b936dc3a-ed20-4041-8bb9-377300c736dc.png">

2. Install snap from the binary files
3. Install microk8s from the binary files
4. Install juju from the binary files
5. Enable microk8s addons (dns, storage, ingress and metallb)

#### Introspection Report
[Uploading inspection-report-20230504_102208.tar.gz…]()

**Comment by neoaggelos** [comment]
Hi @morenoloza98, thank you for reporting the issue.

Can you check that all the Kubernetes services and pods (e.g. for hostpath storage, ingress, dns, etc) are starting properly in the cluster? Running a `microk8s kubectl get all -A` should give you a quick bird's eye view of what is happening.

Note that the ingress, dns, etc services are running as pods inside the cluster, and given that this is an airgap environment, there might image pull failures unless you have properly setup your container registries and/or sideloaded the required images.

The documentation can help you with this, or feel free to ask here if you encounter any unexpected issues.

**Comment by morenoloza98** [comment]
Hi, @neoaggelos. Thank you for taking the time to respond and look into the issue. 

Here is an image of the command ```microk8s kubectl get all -A```

![image](https://user-images.githubusercontent.com/45719935/236838672-26cd1a2e-be08-4692-bbbc-af31b6e5bee1.png)

While enabling the addons I did not get any errors, the only one that timed out was metallb with the following log:

![image](https://user-images.githubusercontent.com/45719935/236840478-f0d48d1b-9d81-4251-9752-421524ef8d03.png)

I hope these give a clearer picture of the issue and thanks again for taking the time.

**Comment by neoaggelos** [comment]
Hi @morenoloza98 

Yes, it looks like the Calico CNI pods (calico-node) are not coming up, thus there is no cni and the rest of the pods are not scheduled at all. This is probably due to the fact that containerd cannot pull the required images.

Looking at the containerd logs with something like `sudo journalctl -u snap.microk8s.daemon-containerd -n 10000` might help, and should include the error messages from containerd failing to pull images.

My suggestion would be to either setup a registry mirror that you control and point containerd to that ( https://microk8s.io/docs/registry-private#configure-registry-mirrors-7), or consider side-loading the required images (https://microk8s.io/docs/sideload).

**Comment by morenoloza98** [comment]
Thanks, again @neoaggelos! I just saved the log to a txt file to analyze it and after doing so I found this concurrent error:

```time="2023-05-08T06:58:17.027761342-07:00" level=error msg="RunPodSandbox for &PodSandboxMetadata{Name:nginx-ingress-microk8s-controller-v44k4,Uid:e780c78e-bb16-412c-a0ce-873f3fcd8c11,Namespace:ingress,Attempt:0,} failed, error" error="rpc error: code = DeadlineExceeded desc = failed to get sandbox image \"registry.k8s.io/pause:3.7\": failed to pull image \"registry.k8s.io/pause:3.7\": failed to pull and unpack image \"registry.k8s.io/pause:3.7\": failed to resolve reference \"registry.k8s.io/pause:3.7\": failed to do request: Head \"https://registry.k8s.io/v2/pause/manifests/3.7\```

So, if I understood correctly, and following the documentation for sideloading, I would have to download this said image (registry.k8s.io/pause:3.7) from a system with internet access and load it to the airgapped system and import it to the cluster. 

Is the link provided above the correct one to pull the image from?

Thanks for answering my questions.

**Comment by neoaggelos** [comment]
Yes, this is correct. FYI, for common addons (ingress, dns, storage, calico) you can see the proper list of images to download in https://github.com/canonical/microk8s/blob/1.26/build-scripts/images.txt

Make sure to change "1.26" in the link above if you are running a different MicroK8s version. Hope this helps!

The documentation should help you with making a bundle of all the required images that you can then import to the airgap environment.

**Comment by morenoloza98** [comment]
Ok, sounds good! Thank you so much for all the help and I will keep you updated on this!

**Comment by morenoloza98** [comment]
Hello again, @neoaggelos! I just wanted to give you a quick update.

After donwloading the images and putting them in the airgapped system, I saw some improvement:

![image](https://user-images.githubusercontent.com/45719935/236933493-9ca94588-a940-470e-a375-e3b3c178bed9.png)

I still need to figure out the `juju bootstrap microk8s` command as that one is still not working, but microk8s' pods are running now.

**Comment by neoaggelos** [comment]
@morenoloza98 What kind of issue are you experiencing with `juju bootstrap microk8s`? bootstrapping a juju controller on the cluster will also require pulling in some images, can you try `juju bootstrap microk8s --debug` and see what the failure message looks like? Also, try to see if you can do `microk8s kubectl get pod -A` while the bootstrap is in progress (to check for any pending/failing to start pods).

**Comment by morenoloza98** [comment]
@neoaggelos I used `juju bootstrap microk8s` and it show the following log

```
Creating Juju controller "microk8s-localhost" on microk8s/localhost
Bootstrap to Kubernetes cluster identified as microk8s/localhost
WARNING Got error requesting "https://streams.canonical.com/juju/gui/streams/v1/index2.sjson": Get "https://streams.canonical.com/juju/gui/streams/v1/index2.sjson": dial tcp 185.125.190.37:443: i/o timeout
WARNING Got error requesting "https://streams.canonical.com/juju/gui/streams/v1/index.sjson": Get "https://streams.canonical.com/juju/gui/streams/v1/index.sjson": dial tcp 185.125.190.37:443: i/o timeout
Unable to fetch Juju Dashboard info: error fetching simplestreams metadata: "https://streams.canonical.com/juju/gui/streams/v1/index.sjson": Get "https://streams.canonical.com/juju/gui/streams/v1/index.sjson": dial tcp 185.125.190.37:443: i/o timeout
Creating k8s resources for controller "controller-microk8s-localhost"
Downloading images
ERROR failed to bootstrap model: creating controller stack: creating statefulset for controller: timed out waiting for controller pod: pending:  - 
```

And it stays in that last error. I understand that juju won't be able to communicate to https://streams.canonical.com... as I have no access to the internet.

Another issue is that juju is creating the controller `microk8s-localhost` and if I try to delete it, it also freezes and shows an error saying 

```
$ juju kill-controller microk8s-localhost 
WARNING! This command will destroy the "microk8s-localhost" controller.
This includes all machines, applications, data and other resources.

Continue? (y/N):y
Unable to open API: no controller API addresses; is bootstrap still in progress?
Unable to connect to the API server, destroying through provider
```
and stays there for a while. It sometimes delete the controller and sometimes does not.

I will run the bootstrap command with the debug flag and post the logs.

Thanks again.

**Comment by morenoloza98** [comment]
This is a great example of the controller behavior I was talking about. There seems to be no controllers (bottom of the image), but it says that there is still a controller named microk8s-localhost

<img width="955" alt="image" src="https://github.com/canonical/microk8s/assets/45719935/61ff8b39-2012-40f9-accb-f079940346f6">

Thanks for your patience and help.

**Comment by neoaggelos** [comment]
I'm not sure how `juju bootstrap microk8s` handles failures, could it be that a previous controller pod was started in the cluster but is simply stuck in a failing state?

Perhaps `juju bootstrap microk8s c1 --debug` can help? (bootstrap a controller with name `c1`, so that there are no conflicts with existing pods)

**Comment by morenoloza98** [comment]
Hi @neoaggelos. There are new updates on the deployment. I have successfully created the controller using `juju bootstrap microk8s`. Right now I am trying to do the kubeflow deployment without internet access/connection.

For further use and reference for anyone experiencing this: 

What I did is that I ran two separate machines, one offline and one online. As I ran the commands in the online machine I reviewed which images were being downloaded with `microk8s ctr image list` and then exported them with `microk8s ctr image export <filename>.tar <image_name>`. I then transferred them to the offline machine and imported them using `microk8s ctr image import - < <filename>.tar`. After that I ran the commands mentioned in the initial comment and everything was running stable.

As I said, right now the next step is not working still as I am trying to figure out how to deploy kubeflow with juju without internet access/connection.
